### PR TITLE
[Feature] Add fissa model

### PR DESCRIPTION
**Pull request by Tokkiu**
Add FISSA model to recbole.
https://doi.org/10.1145/3383313.3412247

### CODE REVIEW

File: recbole/model/sequential_recommender/fissa.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    def calculate_loss(self, interaction):
+        item_seq = interaction[self.ITEM_SEQ]
+        item_seq_len = interaction[self.ITEM_SEQ_LEN]
+        pos_items = interaction[self.POS_ITEM_ID]
+
+        seq_x, seq_y = self.forward(item_seq, item_seq_len)
+        test_item_emb = self.item_embedding.weight
+        seq_output = self.cal_final(seq_x, seq_y, item_seq)
+        logits = torch.matmul(seq_output, test_item_emb.transpose(0, 1))
+        loss = self.loss_fct(logits, pos_items)
```
Comment (Wicknight) [inline recbole/model/sequential_recommender/fissa.py:111]: 这里应该要依据loss_type区分loss的计算

File: recbole/model/sequential_recommender/fissa.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        last_embedding = self.item_embedding(item_seq[:, -1]).to(self.device)  # [B, 1, D]
+        candidates_embedding = self.item_embedding.weight
+        if test_candidates is not None:
+            candidates_embedding = self.item_embedding(test_candidates).to(self.device)  # [B, 1, D]
+        g = torch.cat((last_embedding, seq_y), 1)
+        # g = torch.cat((g, candidates_embedding), 2)
+        g = self.gating_sig(self.gating_lr(g))
+        seq = seq_x * g + seq_y * (1-g)
+        return seq
+
```
Comment (Wicknight) [inline recbole/model/sequential_recommender/fissa.py:124]: 过多空行似乎无意义，需要简单规范一下

File: recbole/model/sequential_recommender/fissa.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        logits = torch.matmul(seq_output, test_item_emb.transpose(0, 1))
+        loss = self.loss_fct(logits, pos_items)
+        return loss
+
+    def cal_final(self, seq_x, seq_y, item_seq, test_candidates=None):
+        last_embedding = self.item_embedding(item_seq[:, -1]).to(self.device)  # [B, 1, D]
+        candidates_embedding = self.item_embedding.weight
+        if test_candidates is not None:
+            candidates_embedding = self.item_embedding(test_candidates).to(self.device)  # [B, 1, D]
+        g = torch.cat((last_embedding, seq_y), 1)
```
Comment (Wicknight) [inline recbole/model/sequential_recommender/fissa.py:119]: 论文中是三部分embedding的聚合，这里先是对global embedding与recent embedding聚合，将candidate设为可选添加，这是因为不添加candidate embedding的效果更好吗？也可以设置一个参数来控制是否将candidate加入聚合以及控制gating_lr的特征维度

**Review by Wicknight — CHANGES_REQUESTED** [review]

**Comment by Tokkiu** [comment]
@Wicknight Thanks for your kind advice. I have taken them into account and committed to a new version. Plz, check it.

**Comment by Tokkiu** [comment]
@Wicknight Thanks for your advice again. 
Now I just tested the FISSA and SASRec models with all default settings on the ml-100k dataset, which means they use almost all similar hyper parameters for a fair comparison.
The final performance is:
* FISSA: ('recall@10', 0.1315), ('mrr@10', 0.0436), ('ndcg@10', 0.0637), ('hit@10', 0.1315), ('precision@10', 0.0131)
* SASRec: ('recall@10', 0.1283), ('mrr@10', 0.0423), ('ndcg@10', 0.062), ('hit@10', 0.1283), ('precision@10', 0.0128)

File: recbole/model/sequential_recommender/fissa.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+        if self.loss_type == 'BPR':
+            self.loss_fct = BPRLoss()
+        elif self.loss_type == 'CE':
+            self.loss_fct = nn.CrossEntropyLoss()
+        else:
+            raise NotImplementedError("Make sure 'loss_type' in ['BPR', 'CE']!")
+
+        self.D = config['hidden_size']
+        self.initializer_range = 0.01
```
Comment (Wicknight) [inline recbole/model/sequential_recommender/fissa.py:60]: 这里应该要读yaml的参数，config['initializer_range']

File: recbole/model/sequential_recommender/fissa.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+from torch.nn.init import xavier_normal_
+
+from recbole.model.abstract_recommender import SequentialRecommender
+from recbole.model.loss import BPRLoss
+from recbole.utils import InputType
+from recbole.model.layers import TransformerEncoder
+
+
+class FISSA(SequentialRecommender):
+    input_type = InputType.PAIRWISE
```
Comment (Wicknight) [inline recbole/model/sequential_recommender/fissa.py:31]: 既然loss_type可选，这里的input_type似乎不需要指定

**Review by Wicknight — CHANGES_REQUESTED** [review]

**Comment by Tokkiu** [comment]
@Wicknight Thanks for your kind comments. I fixed them according to your advice. It makes sense!

**Comment by Tokkiu** [comment]
@Wicknight Thank you for your comments. Is there anything need to be improved?

**Comment by Wicknight** [comment]
> @Wicknight Thank you for your comments. Is there anything need to be improved?

Thank you for your support. I am testing on other datasets and waiting for the results. I will inform you as soon as I finish the test.

**Comment by Wicknight** [comment]
@Tokkiu During the test, I had a little doubt about the candidate items in line 116 and line 117. All items are used as candidate items in the code here. This increases the burden on the GPU and it seems to be inconsistent with the original intention of the paper. Is there any alternative in your opinion?

**Comment by Tokkiu** [comment]
> 

According to the paper, they adopt binary cross entropy. Whereas we usually use CE and BPR in recbole for most sequential models. I also think it's improper to use all items when calculating loss frequently. So from me, a viable way to solve this is use BCE + sampling instead of full CE. How do you think?

**Comment by Wicknight** [comment]
> @Wicknight Thanks for your concern. According to the paper, they adopt binary cross-entropy. Whereas we usually use CE and BPR in recbole for most sequential models. I also think it's improper to forward all items when calculating loss frequently. So from me, a viable way to solve this is using BCE + sampling instead of full CE. How do you think?

@Tokkiu I think this is a good suggestion. Thank you for your help again! I have conducted certain tests, but it seems that I can not reproduce the results reported in the paper. Therefore, I would like to ask whether you have successfully reproduced the results of the paper？

**Comment by Tokkiu** [comment]
> > @Wicknight Thanks for your concern. According to the paper, they adopt binary cross-entropy. Whereas we usually use CE and BPR in recbole for most sequential models. I also think it's improper to forward all items when calculating loss frequently. So from me, a viable way to solve this is using BCE + sampling instead of full CE. How do you think?
> 
> @Tokkiu I think this is a good suggestion. Thank you for your help again! I have conducted certain tests, but it seems that I can not reproduce the results reported in the paper. Therefore, I would like to ask whether you have successfully reproduced the results of the paper？

Yes, I have same conclusion. From my test results. The performance of FISSA is comparable with SASRec. But it can't outperform SASRec consistently as mentioned in paper.

**Comment by Wicknight** [comment]
> > > @Wicknight Thanks for your concern. According to the paper, they adopt binary cross-entropy. Whereas we usually use CE and BPR in recbole for most sequential models. I also think it's improper to forward all items when calculating loss frequently. So from me, a viable way to solve this is using BCE + sampling instead of full CE. How do you think?
> > 
> > 
> > @Tokkiu I think this is a good suggestion. Thank you for your help again! I have conducted certain tests, but it seems that I can not reproduce the results reported in the paper. Therefore, I would like to ask whether you have successfully reproduced the results of the paper？
> 
> Yes, I have same conclusion. From my test results. The performance of FISSA is comparable with SASRec. But it can't outperform SASRec consistently as mentioned in paper.

Could you tell me how many datasets in the paper have you used for testing？Does this inconsistency occur on all datasets？

**Comment by Tokkiu** [comment]
> > > > @Wicknight Thanks for your concern. According to the paper, they adopt binary cross-entropy. Whereas we usually use CE and BPR in recbole for most sequential models. I also think it's improper to forward all items when calculating loss frequently. So from me, a viable way to solve this is using BCE + sampling instead of full CE. How do you think?
> > > 
> > > 
> > > @Tokkiu I think this is a good suggestion. Thank you for your help again! I have conducted certain tests, but it seems that I can not reproduce the results reported in the paper. Therefore, I would like to ask whether you have successfully reproduced the results of the paper？
> > 
> > 
> > Yes, I have same conclusion. From my test results. The performance of FISSA is comparable with SASRec. But it can't outperform SASRec consistently as mentioned in paper.
> 
> Could you tell me how many datasets in the paper have you used for testing？Does this inconsistency occur on all datasets？

I mainly test datasets in MovieLens. It shows similar performance on ml-1m and ml-10m with comparable performance on ml-100k.

**Comment by Wicknight** [comment]
> > > > > @Wicknight Thanks for your concern. According to the paper, they adopt binary cross-entropy. Whereas we usually use CE and BPR in recbole for most sequential models. I also think it's improper to forward all items when calculating loss frequently. So from me, a viable way to solve this is using BCE + sampling instead of full CE. How do you think?
> > > > 
> > > > 
> > > > @Tokkiu I think this is a good suggestion. Thank you for your help again! I have conducted certain tests, but it seems that I can not reproduce the results reported in the paper. Therefore, I would like to ask whether you have successfully reproduced the results of the paper？
> > > 
> > > 
> > > Yes, I have same conclusion. From my test results. The performance of FISSA is comparable with SASRec. But it can't outperform SASRec consistently as mentioned in paper.
> > 
> > 
> > Could you tell me how many datasets in the paper have you used for testing？Does this inconsistency occur on all datasets？
> 
> I mainly test datasets in MovieLens. It shows similar performance on ml-1m and ml-10m with comparable performance on ml-100k.

So you didn't test on the five datasets used in the paper？

**Comment by Tokkiu** [comment]
> 
Yes, did you test on them？How about performance gap？

**Comment by Tokkiu** [comment]
> > > 
> > 
> > 
> > Yes, did you test on them？How about performance gap？
> 
> First of all, thank you for your PR and efforts to RecBole! I didn't test on all datasets, but the inconsistency occurred on several datasets. So If you want to launch this model, you must be able to reproduce the results of the paper. Thank you again for your contribution! I also hope you can continue to discuss the reasons for this inconsistency with me.

Nice suggestion! I am executing experiments on datasets mainly used in the paper. And I think the inconsistency may come from hyperparameter settings, e.g., max sequence length, dropout ratio, and negative sampling.
